extensions/objects/templates/unicode-string-editor.component.spec.ts

```diff
@@ -0,0 +1,164 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Unit tests for unicode string editor.
+ */
+
+import { FormsModule } from '@angular/forms';
+import { EventEmitter, SimpleChanges } from '@angular/core';
+import { async, ComponentFixture, TestBed } from '@angular/core/testing';
+import { UnicodeStringEditorComponent } from './unicode-string-editor.component';
+import { importAllAngularServices } from 'tests/unit-test-utils';
+import { ExternalSaveService } from 'services/external-save.service';
+
+describe('UnicodeStringEditorComponent', () => {
+  let component: UnicodeStringEditorComponent;
+  let fixture: ComponentFixture<UnicodeStringEditorComponent>;
+  let externalSaveService: ExternalSaveService;
+
+  var externalSaveEventEmitter = new EventEmitter();
+
+  beforeEach(async(() => {
+    TestBed.configureTestingModule({
+      imports: [FormsModule],
+      declarations: [UnicodeStringEditorComponent]
+    }).compileComponents();
+  }));
+
+  importAllAngularServices();
+
+  beforeEach(angular.mock.module('oppia'));
+  beforeEach(() => {
+    fixture = TestBed.createComponent(UnicodeStringEditorComponent);
+    component = fixture.componentInstance;
+    externalSaveService = TestBed.get(ExternalSaveService);
+    spyOnProperty(externalSaveService, 'onExternalSave').and.returnValue(
+      externalSaveEventEmitter);
+    component.alwaysEditable = false;
+    component.value = 'random value';
+  });
+
+  afterEach(() => {
+    component.ngOnDestroy();
+  });
+
+  it('should initialize component when interaction rule is selected', () => {
+    spyOn(component, 'closeEditor').and.callThrough();
+    spyOn(component.componentSubscriptions, 'add');
+    component.active = true;
+
+    expect(component.active).toBe(true);
+
+    component.ngOnInit();
+
+    // We cannot test what componentSubscriptions.add has been called with since
+    // externalSaveService is a private variable and inside it is an anonymous
+    // fucntion.
+    expect(component.componentSubscriptions.add).toHaveBeenCalled();
+    expect(component.componentSubscriptions.closed).toBe(false);
+    expect(component.alwaysEditable).toBe(false);
+    expect(component.closeEditor).toHaveBeenCalled();
+    expect(component.active).toBe(false);
+  });
+
+  it('should replace value when External Save is emitted', () => {
+    component.active = true;
+    component.ngOnInit();
+    spyOn(component, 'closeEditor').and.callThrough();
+    spyOn(component, 'replaceValue').and.callThrough();
+    component.active = true;
+
+    expect(component.active).toBe(true);
+
+    externalSaveEventEmitter.emit();
+
+    // The this.value cannot be tested for changes since it set before the
```

Comment:
Hmm, what I don't understand is the purpose of the event and the code....

Reply:
No, what I mean is what does the event achieve? It doesn't seem to assign anything new

Reply:
Same question as @nithusha21 from my end. A test needs to be testing a change; if you can't test that then the test can't claim to be testing that. So I don't understand what's going on either.

Reply:
Then why not update the code in line 46 to be this.closeEditor() instead, if some of it is a no-op? Writing tests for code often results in improving it if you find things that don't logically make sense.

(Please still manually test to make sure it works correctly.)

Reply:
removed